From 4092fd62a82c35ee1659626bcc88597b47423aef Mon Sep 17 00:00:00 2001
From: Stephane Viau <stephane.viau@nxp.com>
Date: Mon, 26 Apr 2021 15:48:20 +0200
Subject: [PATCH 04/43] HRPN-38 imx8mp-freertos: New cell

The -rtos header can be reused for multiple RTOS (such as upcoming Zephyr)
cell configurations ;
the -freertos configuration file is used for the industrial-flavor of the
Harpoon project.

Signed-off-by: Stephane Viau <stephane.viau@nxp.com>
---
 configs/arm64/imx8mp-freertos.c | 45 +++++++++++++++++++++++++++++++++
 configs/arm64/imx8mp-rtos.h     | 38 ++++++++++++++++++++++++++++
 2 files changed, 83 insertions(+)
 create mode 100644 configs/arm64/imx8mp-freertos.c
 create mode 100644 configs/arm64/imx8mp-rtos.h

diff --git a/configs/arm64/imx8mp-freertos.c b/configs/arm64/imx8mp-freertos.c
new file mode 100644
index 00000000..96f757d1
--- /dev/null
+++ b/configs/arm64/imx8mp-freertos.c
@@ -0,0 +1,45 @@
+/*
+ * i.MX8MP target - FreeRTOS
+ *
+ * Copyright 2021 NXP
+ *
+ * Authors:
+ *  Stephane Viau <stephane.viau@nxp.com>
+ *  Fabrice Goucem <fabrice.goucem@nxp.com>
+ *
+ * This work is licensed under the terms of the GNU GPL, version 2.  See
+ * the COPYING file in the top-level directory.
+ */
+
+#include "cell-helper.h"
+#include "imx8mp-rtos.h"
+
+/* Name, cores, entry point */
+#define CONFIG_INMATE_NAME		"freertos"
+#define CONFIG_INMATE_CORE_BITMAP	(0b1000)
+#define CONFIG_INMATE_BASE		(0xc0000000)
+
+/* Memory & peripherals */
+#define CONFIG_INMATE_REGIONS_NUM	(RTOS_COMMON_MEMORY_REGIONS_NUM + 4)
+#define CONFIG_INMATE_REGIONS		\
+	RTOS_COMMON_MEMORY_REGIONS,                                     \
+	MEM_REGION_RWXL(0xc0000000, 0xc0000000, MB(16)),   /* RAM */    \
+	MEM_REGION_RWX( 0x00900000, 0x00900000, KB(384)),  /* OCRAM */  \
+	MEM_REGION_RW(  0x00800000, 0x00800000, KB(128)),  /* DTCM */   \
+	MEM_REGION_RWX( 0x007e0000, 0x007e0000, KB(128 )), /* ITCM */   \
+
+/* GIC */
+#define CONFIG_INMATE_IRQCHIPS_NUM	(RTOS_COMMON_IRQCHIPS_NUM)
+#define CONFIG_INMATE_IRQCHIPS_ADDR	(0x38800000)
+#define CONFIG_INMATE_IRQCHIPS_BASE	(32)
+#define CONFIG_INMATE_IRQCHIPS_BITMAP	\
+	/* interrupts 32..63 */         \
+	RTOS_COMMON_IRQCHIPS_BITMAP1,   \
+	/* interrupts 64..95 */         \
+	RTOS_COMMON_IRQCHIPS_BITMAP2,   \
+	/* interrupts 96..127 */        \
+	RTOS_COMMON_IRQCHIPS_BITMAP3,   \
+	/* interrupts 128..159 */       \
+	RTOS_COMMON_IRQCHIPS_BITMAP4    \
+
+#include "cell-create.h"
diff --git a/configs/arm64/imx8mp-rtos.h b/configs/arm64/imx8mp-rtos.h
new file mode 100644
index 00000000..a7f13350
--- /dev/null
+++ b/configs/arm64/imx8mp-rtos.h
@@ -0,0 +1,38 @@
+/*
+ * i.MX8MP target - common RTOS config
+ *
+ * Copyright 2021 NXP
+ *
+ * Authors:
+ *  Stephane Viau <stephane.viau@nxp.com>
+ *
+ * This work is licensed under the terms of the GNU GPL, version 2.  See
+ * the COPYING file in the top-level directory.
+ *
+ */
+
+#define RTOS_COMMON_MEMORY_REGIONS_NUM                       5
+#define RTOS_COMMON_MEMORY_REGIONS \
+	MMIO_REGION_RW( 0x302d0000, 0x302d0000, KB(64)), /* GPT1 */                 \
+	MMIO_REGION_RW( 0x302e0000, 0x302e0000, KB(64)), /* GPT2 */                 \
+	MMIO_REGION_ROS(0x30360000, 0x30360000, KB(64)), /* ANA_PLL */              \
+	MMIO_REGION_ROS(0x30380000, 0x30380000, KB(64)), /* CCM */                  \
+	MMIO_REGION_RW( 0x30a60000, 0x30a60000, KB(64))  /* UART4 */
+
+#define RTOS_COMMON_IRQCHIPS_NUM                             1
+/* interrupts 32..63 */
+#define RTOS_COMMON_IRQCHIPS_BITMAP1 \
+	(1 << (29 + 32 - 32)) /* UART4 */
+
+/* interrupts 64..95 */
+#define RTOS_COMMON_IRQCHIPS_BITMAP2 \
+	(1 << (54 + 32 - 64)) | /* GPT2 */ \
+	(1 << (55 + 32 - 64))   /* GPT1 */
+
+/* interrupts 96..127 */
+#define RTOS_COMMON_IRQCHIPS_BITMAP3 \
+	0
+
+/* interrupts 128..159 */
+#define RTOS_COMMON_IRQCHIPS_BITMAP4 \
+	0
-- 
2.25.1

